crypto/fipsmodule/kdf/sskdf.c

```diff
@@ -0,0 +1,314 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+#include <assert.h>
+#include <openssl/base.h>
+#include <openssl/digest.h>
+#include <openssl/hmac.h>
+#include <openssl/kdf.h>
+#include "../delocate.h"
+#include "internal.h"
+
+static int SSKDF_VARIANT_DIGEST_CTX_init(SSKDF_VARIANT_CTX *ctx,
+                                         const EVP_MD *digest) {
+  SSKDF_VARIANT_DIGEST_CTX *variant_ctx = NULL;
+  EVP_MD_CTX *md_ctx = NULL;
+
+  int ret = 0;
+
+  if (!ctx || !digest) {
+    goto err;
+  }
+
+  variant_ctx = OPENSSL_malloc(sizeof(SSKDF_VARIANT_DIGEST_CTX));
+  if (!variant_ctx) {
+    goto err;
+  }
+
+  md_ctx = EVP_MD_CTX_new();
+  if (!md_ctx) {
+    goto err;
+  }
+
+  ret = 1;
+  variant_ctx->digest = digest;
+  variant_ctx->ctx = md_ctx;
+  ctx->data = variant_ctx;
+
+  return ret;
+
+err:
+  EVP_MD_CTX_free(md_ctx);
+  OPENSSL_free(variant_ctx);
+
+  return ret;
+}
+
+
+
+static void SSKDF_VARIANT_DIGEST_CTX_cleanup(SSKDF_VARIANT_CTX *ctx) {
+  if (!ctx) {
+    return;
+  }
+  SSKDF_VARIANT_DIGEST_CTX *variant_ctx = (SSKDF_VARIANT_DIGEST_CTX *)ctx->data;
+  if (!ctx->data) {
+    return;
+  }
+  EVP_MD_CTX_free(variant_ctx->ctx);
+  OPENSSL_free(variant_ctx);
+  ctx->data = NULL;
+}
+
+static int SSKDF_VARIANT_HMAC_CTX_init(SSKDF_VARIANT_CTX *ctx,
+                                       const EVP_MD *digest,
+                                       const uint8_t *salt, size_t salt_len) {
+  SSKDF_VARIANT_HMAC_CTX *variant_ctx = NULL;
+  HMAC_CTX *hmac_ctx = NULL;
+
+  int ret = 0;
+
+  if (!ctx || !digest) {
+    goto err;
+  }
+
+  variant_ctx = OPENSSL_malloc(sizeof(SSKDF_VARIANT_HMAC_CTX));
+  if (!variant_ctx) {
+    goto err;
+  }
+
+  hmac_ctx = HMAC_CTX_new();
+  if (!hmac_ctx) {
+    goto err;
+  }
+
+  if (!HMAC_Init_ex(hmac_ctx, salt, salt_len, digest, NULL)) {
+    goto err;
+  }
+
+  ret = 1;
+  variant_ctx->ctx = hmac_ctx;
+  ctx->data = variant_ctx;
+
+  return ret;
+
+err:
+  HMAC_CTX_free(hmac_ctx);
+  OPENSSL_free(variant_ctx);
+
+  return ret;
+}
+
+
+
+static void SSKDF_VARIANT_HMAC_CTX_cleanup(SSKDF_VARIANT_CTX *ctx) {
+  if (!ctx) {
+    return;
+  }
+  SSKDF_VARIANT_HMAC_CTX *hmac_ctx = (SSKDF_VARIANT_HMAC_CTX *)ctx->data;
+  if (!hmac_ctx) {
+    return;
+  }
+  HMAC_CTX_free(hmac_ctx->ctx);
+  OPENSSL_free(hmac_ctx);
+  ctx->data = NULL;
```

Comment:
Nit:  I would consider just a memset of the structure.  Its effectively the same thing but easier to read.